private/pkg/storage/matcher.go

```diff
@@ -26,6 +26,13 @@ type Matcher interface {
 	//
 	// The path is expected to be normalized and validated.
 	MatchPath(string) bool
+	// MatchPrefix returns the matching prefixes.
+	//
+	// The prefix is expected to be normalized and validated.
+	// If the prefix is only partially matched, the valid child prefixes are returned.
+	// If the prefix is fully matched, the prefix is returned.
+	// If the prefix is not matched, nil is returned.
+	MatchPrefix(string) []string
```

Comment:
If `MatchPath` returns `bool`, I would expect all `Match.*` functions to return `bool`. A different naming convention will need to  be used.

I don't understand how this interacts with `MatchPath`. Is it valid to have a situation where a value returned from `MatchPrefix` would return `false` for `MatchPatch`? Is this validated in some way, and are the creation of `Matchers` restricted to this package/should they be? Why or why not would a caller outside of this package need `MatchPrefix`? Is the returned slice a copy, or should it not be mutated?

This PR is editing a foundational interfact, so all these questions need to be answered as cleanly as possible, and documented.

Reply:
Will fix naming.

`MatchPrefix` returns valid child prefixes. If a bucket is constructed with `MatchPathEqualOrContained("a/b/c")` and the caller walks at `.` we want to ignore all paths outside `a/b/c` so should only walk at the prefix `a/b/c`. In this case `MatchPath(".")` would return false. If we walk `a/b/c/d` we have a parent include so want `MatchPrefix("a/b/c/d")` to return the same path `a/b/c/d` and `MatchPath("a/b/c/d")` would return true.

`Matchers` are restricted to this package with the `isMatchers` method, this is the same as `Mappers` with the `isMappers`. Callers outside wouldn't need either `MatchPath` or `MatchPrefix`, from grepping I don't think any code uses Mappers to use `MapPath` or `UnmapFullPath` outside of this package. Returned slice is a copy, can be mutated.